### ISSUE TITLE
Getting "Problem loading key" error on test server with LWS + mbedtls build

### ISSUE THREAD

**Issue by rpati12** [warmcat/libwebsockets#2992]
I am trying to run lws test server with mbedtls on Linux and getting following logs:

> /libwebsockets-test-server --port=4433 --ssl
> [2023/10/31 08:52:35:5659] N: libwebsockets test server - license MIT
> [2023/10/31 08:52:35:5661] N: (C) Copyright 2010-2018 Andy Green <andy@warmcat.com>
> Using resource path "/home/projects/lws/libwebsockets/_build/install/share/libwebsockets-test-server"
> [2023/10/31 08:52:35:5661] N: lws_create_context: LWS: 4.3.99-v4.3.0-282-g18fdb0d6, NET CLI SRV H1 H2 WS SS-JSON-POL MbedTLS ConMon IPv6-absent
> [2023/10/31 08:52:35:5663] N: __lws_lc_tag:  ++ [wsi|0|pipe] (1)
> [2023/10/31 08:52:35:5664] N: __lws_lc_tag:  ++ [vh|0|netlink] (1)
> [2023/10/31 08:52:35:5668] N: __lws_lc_tag:  ++ [vh|1|_ss_default||-1] (2)
> [2023/10/31 08:52:35:5668] N: __lws_lc_tag:  ++ [vh|2|default||4433] (3)
> [2023/10/31 08:52:35:5668] N:  SSL ciphers: 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4:!HMAC_SHA1:!SHA1:!DHE-RSA-AES128-GCM-SHA256:!DHE-R...
> [2023/10/31 08:52:35:5668] N:  Vhost 'default' using TLS mode
> [2023/10/31 08:52:35:5668] N: lws_tls_server_vhost_backend_init: vh default: mem CA OK
> [2023/10/31 08:52:35:5671] E: Problem loading key
> [2023/10/31 08:52:35:5671] E: [vh|2|default||4433]: lws_create_vhost: lws_context_init_server_ssl failed
> [2023/10/31 08:52:35:5671] N: __lws_lc_untag:  -- [vh|2|default||4433] (2) 282μs
> [2023/10/31 08:52:35:5671] E: vhost creation failed

It does not start with above error. Is that supported ? Could anyone point out what I am missing ?

**Comment by lws-team** [comment]
What mbedtls version is it and what certs are you feeding it?  The error is coming when it tries to extract the private key from the server private cert.  You might get more info building in debug mode and using -d2047 or such.

**Comment by rpati12** [comment]
I have pulled the latest mbedtls branch though.
I am using `libwebsockets-test-server` application, so I think it is using following vert and key files:
```
libwebsockets-test-server.pem
libwebsockets-test-server.key.pem
```

I have just copied the files into the binary directory.
When I use the mbedtls `key_app` application to read the key file, it is successful with following logs:
./key_app mode=private filename=libwebsockets-test-server.key.pem

>   . Loading the private key ... ok
>   . Key information    ...
> N:  C99C95F6A45DD09D0ABCC71E0E62001B9F2C81D1A7332B037E9A171EE6CCF084DE369BE733EEE475E491FADCA4C62C60E50B45CFED87C8C4D59FF3AA50D51CAA61755D5BD675660EA9B5C8296377D6BDBAC1CE7710CAF5A2DC0C46AD3E6FDD086B368A2FF7288D675B3387BC3C954E443B78FDB6665DCEA1387108B27C04CD38DA1933A417D85026CCF08B3A88162B0E18C83120C52AE190B5ED5F68C622A78B4DDCD10EFFF5CFEDD2ACA19B64E933224AB26A766D1534B3943B8AA41DDF4975CC76254D4AAC2550FA32725634A3E320153B0C4F830C78E85F4B0FBF21F8F70D1F663C4153B6A9E4ED583D69E4E263D37D7C4AC33DE6F4E89A6CC8A6E6D5B517
> E:  010001
> D:  991DE832DA6039A89B5805ED88C4DB0A1F5860FBC5FD3164C508E8185AF7388FE0681EFD77CEAE09EF5F02E401EC5415FEBE11F63992CF62D64F55727B536F22E3438C74CC2A7B2915B0EE64A89B8226B428AED274997FF681227DA6E36D77B40620F4D03AE8BC83CB9B16325ADC9605721990AEAF8DFCF92A7186466FB401DB19CAE33A0D661C9198ACDA645FDEF54B26ED3F6957BE91F9D87FFA0EE24A508B3533BC11269BA11A59360BC18A8C5314795BE4D4057DCAA70D277BD9D2A56A50A05785698189572868721103F608F8DC82BAF4B61791538B524C79A4CA745D92156646C195E8C1072CE973D33946AD13328856B00DCFB792D59D1E1D6DF188F9
> P:  FB46A927B5BD43D58FE5A5A4E24E4BF09577BF9C1618D875998320633324FF15B517DD910BB69FBF5BAF2085DBD1BC86E415B5CD2874168E8B53D1DF1EC45121A602A371FC0F99C4824A6CFE62A8EDECF23596BFDDB7565B4D0D7DD417DB5FC3A67477E5863103993334D896FB6EA4AF0632BD730D78456921BBA80CE0045FF3
> Q:  CD66E69BD634A6E1B572335F15E7721B4642B6CE75941A20D545EA734BA57B35EB11822F4090A3C9E9227599E497B1B092E10A0FD47AC4A8042F9AEE6DC4B9FDB7CBE05D97F38A93F197264BDDE877CEB9AAEC7700AA80F434DF2A017185983340D963018748608EB6E3BB15AD5F880ECC8E6112C122F690E026911EB447034D
> DP: DEC8E58D59AB370B908BB1F3A0DA7EA60E3F9FD94A163944FB18B18C827A18D82C39BEAEAD4DFC10F3BF2FFCBECB14761C9FCB12DCFD204ABDE951389BD4BF822FE842F5FADFF21BB2D13E15C56FB593D1A405C69353CBF3233C05061928DD9D6065FF2FB76C999E29327F9E4DCAE9A46CC06C0C28332C876BEB2FB24AA45655
> DQ:  50C9E5706BFB3F9C4D52C2B3E698A6FFE54DBE3B4C71D9290AD9BA1269318859B7C0A634D1AF524CCB5EDB64914DBCFD17785E0DC5E417CF81FDFB71464977623CD57AC5354BFC389E707606A71C956B010A2281D808383A5545656F35E876E9C9DAF71F15C7B8FAC9229F66324FFDE2DC40301C414CF0302189E9D06AD5F78D
> QP:  DDDEC95C646E738D22A9270820FDD98B9A8CD10AF1C5871023F2858EE13A6A501129EB3ED0DECC54D612527D4D0FFF9925736A14270A6A87DD8D78A0A7074DCB4C329D6EB528C2CF01FD2E3C74F3079F7DF5674D56FA3BD71DB09611F1AF493AECE6A74AC37BA730EDD2CA14610AD8B048CCAB2171054EF931E84A3690EF21C4

**Comment by lws-team** [comment]
> I have pulled the latest mbedtls branch though.

It sounds like you expect that to be the safest bet, but the apis for the tls libraries churn a lot.  What is the exact HEAD commit you are using for mbedtls and what steps did you take to build it?

The logs suggestion may point the finger more closely to what is actually happening.

**Comment by rpati12** [comment]
I have currently the following mbedtls commit:

> https://github.com/Mbed-TLS/mbedtls/tree/28b56335bb3f9073d8bebba752e7b8f5d960f39d

Building mbedtls as:

> cmake -DCMAKE_INSTALL_PREFIX=install -DMBEDTLS_SSL_PROTO_TLS1_3=ON ..
> make install

The debug logs from LWS:

> ./libwebsockets-test-server --port=4433 --ssl -d2047
> [2023/10/31 09:20:51:1483] N: libwebsockets test server - license MIT
> [2023/10/31 09:20:51:1484] N: (C) Copyright 2010-2018 Andy Green <andy@warmcat.com>
> Using resource path "/home/projects/lws/libwebsockets/_build/install/share/libwebsockets-test-server"
> [2023/10/31 09:20:51:1484] D: _realloc: size 9032: context
> [2023/10/31 09:20:51:1485] N: lws_create_context: LWS: 4.3.99-v4.3.0-282-g18fdb0d6, NET CLI SRV H1 H2 WS SS-JSON-POL MbedTLS ConMon IPv6-absent
> [2023/10/31 09:20:51:1485] I: lws_create_context: Event loop: poll
> [2023/10/31 09:20:51:1485] D: _realloc: size 72: lws_smd_register
> [2023/10/31 09:20:51:1485] I: lws_smd_register: peer 0x55c214874930 (count 1) registered
> [2023/10/31 09:20:51:1485] D: _realloc: size 8192: fds table
> [2023/10/31 09:20:51:1485] I: lws_create_context: ctx:  7800B (3704 ctx + pt(1 thr x 4096)), pt-fds: 1024
> [2023/10/31 09:20:51:1485] I: lws_create_context:  http: ah_data: 4096, ah: 992, max count 1024
> [2023/10/31 09:20:51:1485] D: _realloc: size 8192: lws_lookup
> [2023/10/31 09:20:51:1486] I: lws_plat_init:  mem: platform fd map:  8192 B
> [2023/10/31 09:20:51:1486] D: _realloc: size 1232: __lws_wsi_create_with_role
> [2023/10/31 09:20:51:1486] D: : lws_role_transition: wsistate 0x200, ops pipe
> [2023/10/31 09:20:51:1486] N: __lws_lc_tag:  ++ [wsi|0|pipe] (1)
> [2023/10/31 09:20:51:1486] D: [wsi|0|pipe]: __insert_wsi_socket_into_fds: tsi=0, sock=4, pos-in-fds=0
> [2023/10/31 09:20:51:1486] I: rops_pt_init_destroy_netlink: creating netlink skt
> [2023/10/31 09:20:51:1486] D: _realloc: size 1232: __lws_wsi_create_with_role
> [2023/10/31 09:20:51:1486] D: : lws_role_transition: wsistate 0x200, ops netlink
> [2023/10/31 09:20:51:1487] N: __lws_lc_tag:  ++ [vh|0|netlink] (1)
> [2023/10/31 09:20:51:1487] D: [vh|0|netlink]: __insert_wsi_socket_into_fds: tsi=0, sock=5, pos-in-fds=1
> [2023/10/31 09:20:51:1487] D: rops_pt_init_destroy_netlink: starting netlink coldplug wait
> [2023/10/31 09:20:51:1487] I:  Compiled with MbedTLS support
> [2023/10/31 09:20:51:1487] I: lws_server_get_canonical_hostname:  canonical_hostname = dev
> [2023/10/31 09:20:51:1487] D: _realloc: size 1128: lws_ss_policy_parse_begin
> [2023/10/31 09:20:51:1489] I: lws_ss_policy_set: 2.064KiB, pad 56%: hardcoded
> [2023/10/31 09:20:51:1489] D: _realloc: size 916: lws_create_vhost
> [2023/10/31 09:20:51:1489] N: __lws_lc_tag:  ++ [vh|1|_ss_default||-1] (2)
> [2023/10/31 09:20:51:1489] D: _realloc: size 560: vh plugin table
> [2023/10/31 09:20:51:1489] D: _realloc: size 216: same vh list
> [2023/10/31 09:20:51:1489] I: [vh|1|_ss_default||-1]: lws_create_vhost: Creating Vhost '_ss_default' (serving disabled), 9 protocols, IPv6 off
> [2023/10/31 09:20:51:1489] I: created client ssl context for _ss_default
> [2023/10/31 09:20:51:1489] I: lws_create_context:  mem: per-conn:         1232 bytes + protocol rx buf
> [2023/10/31 09:20:51:1489] D: lws_cancel_service: 
> [2023/10/31 09:20:51:1489] D: _realloc: size 912: lws_create_vhost
> [2023/10/31 09:20:51:1489] N: __lws_lc_tag:  ++ [vh|2|default||4433] (3)
> [2023/10/31 09:20:51:1489] D: _realloc: size 224: vh paths
> [2023/10/31 09:20:51:1489] D: _realloc: size 560: vh plugin table
> [2023/10/31 09:20:51:1489] D: _realloc: size 216: same vh list
> [2023/10/31 09:20:51:1489] I: [vh|2|default||4433]: lws_create_vhost: Creating Vhost 'default' port 4433, 9 protocols, IPv6 off
> [2023/10/31 09:20:51:1489] I: [vh|2|default||4433]: lws_create_vhost:    mounting file:///home/projects/lws/libwebsockets/_build/install/share/libwebsockets-test-server to /
> [2023/10/31 09:20:51:1489] I: [vh|2|default||4433]: lws_create_vhost:    mounting callback://protocol-post-demo to /formtest
> [2023/10/31 09:20:51:1489] I: [vh|2|default||4433]: lws_create_vhost:    mounting file:///home/projects/lws/libwebsockets/_build/install/share/libwebsockets-test-server/candide.zip to /ziptest
> [2023/10/31 09:20:51:1489] I: [vh|2|default||4433]: lws_create_vhost:    mounting file:///home/projects/lws/libwebsockets/_build/install/share/libwebsockets-test-server/candide-uncompressed.zip to /uncommziptest
> [2023/10/31 09:20:51:1489] N:  SSL ciphers: 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4:!HMAC_SHA1:!SHA1:!DHE-RSA-AES128-GCM-SHA256:!DHE-R...
> [2023/10/31 09:20:51:1489] N:  Vhost 'default' using TLS mode
> [2023/10/31 09:20:51:1489] N: lws_tls_server_vhost_backend_init: vh default: mem CA OK
> [2023/10/31 09:20:51:1490] D: _realloc: size 1429: alloc_file
> [2023/10/31 09:20:51:1491] D: _realloc: size 1709: alloc_file
> [2023/10/31 09:20:51:1492] E: Problem loading key
> [2023/10/31 09:20:51:1492] E: [vh|2|default||4433]: lws_create_vhost: lws_context_init_server_ssl failed
> [2023/10/31 09:20:51:1492] I: [vh|2|default||4433]: lws_vhost_destroy1: 
> [2023/10/31 09:20:51:1492] I: [vh|2|default||4433]: lws_vhost_destroy: count_bound_wsi 0
> [2023/10/31 09:20:51:1492] N: __lws_lc_untag:  -- [vh|2|default||4433] (2) 254μs
> [2023/10/31 09:20:51:1492] E: vhost creation failed

**Comment by rpati12** [comment]
This is how I am building LWS:

> cmake -DLWS_WITH_MBEDTLS=1 -DLWS_HAVE_MBEDTLS_NET_SOCKETS=ON -DLWS_MBEDTLS_LIBRARIES=/home/projects/lws/mbedtls/_build/install/lib/libmbedtls.a  -DLWS_MBEDTLS_INCLUDE_DIRS=/home/projects/lws/mbedtls/_build/install/include -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=install -DLWS_WITH_STATIC=ON -DLWS_WITH_SHARED=OFF ..

Note: `libmbedtls.a` is a combined library including crypto and all others from mbedtls. This is probably another issue that I can not provide multiple libraries in command line option `LWS_MBEDTLS_LIBRARIES` or at least I am not aware of how to do that.

**Comment by lws-team** [comment]
I met some deviations following this.

1) I used this to configure mbedtls cmake

```
cmake .. -D-DUSE_SHARED_MBEDTLS_LIBRARY=On
```

the suggested one dumped

```
CMake Warning:
  Manually-specified variables were not used by the project:

    MBEDTLS_SSL_PROTO_TLS1_3
```

2) mbedtls changed the name of a session-related constant. I added a patch to lws to adapt to it.

https://github.com/warmcat/libwebsockets/commit/e71398c02ac2db15e28f38367491af8600867bca

3) In addition the mbedtls installs into build/install, after sudo make install I had to copy the contents of that dir to /usr, that's new

```
sudo make install && sudo cp -rp install/* /usr
```

4) For lws,it wanted

```
 cmake .. -DLWS_WITH_MBEDTLS=1 -DLWS_STATIC_PIC=1 -DLWS_LINK_TESTAPPS_DYNAMIC=1
```

5) After making and make installing lws, I can reproduce it, I'm looking at it.

wrt libs, Cmake usually wants semicolons in lists.

**Comment by rpati12** [comment]
1. MBEDTLS_SSL_PROTO_TLS1_3 -> you are right, I had to turn this on in `mbedtls_config.h` file. May be I had tried to turn it on using cmake but failed.
2. Session constants, yes, I had made same fixes on my end to get it built
3. Install -> I preffered a local install as I am just trying to run it on android at first, and before that I am trying to do it on linux without impacting my existing setup for other projects.
4. LWS build -> is the LWS_HAVE_MBEDTLS_NET_SOCKETS option not required ?
5. Thanks for the hint. That is helpful.

**Comment by rpati12** [comment]
Just a hint if that helps you:

I think the loading should look like this:
![image](https://github.com/warmcat/libwebsockets/assets/141009171/d999b343-ca59-49bc-8354-a88ba56d87b1)

But, in LWS it looks like following(You can ignore my changes, I was just trying to make it similar, but it seems that some other methods are already implemented in ssl_pm to do this, but not getting called):
![image](https://github.com/warmcat/libwebsockets/assets/141009171/76a5e72f-8a3c-4509-8d26-b10425278233)

**Comment by lws-team** [comment]
Lws has worked with mbedtls for many years.  This is to do with breaking changes in it as they make new versions.  It's not to do with making it work for the first time.

If you eg, `git reset --hard mbedtls-2.9.0` and remake mbedtls after `rm CMakeCache.txt ; make clean` and redo lws the same against the older mbedtls, then the whole thing works fine.  It's ONLY that you are using newer mbedtls.  I didn't check when it broke but you can try newer 3.0.0+ tagged versions the same way.

LWS_HAVE_MBEDTLS_NET_SOCKETS is a previous breaking change mbedtls did and lws is detecting it at cmake-time so it can work with both newer and older versions of mbedtls.

Previous mbedtls didn't support v1.3, I'm guessing it's to do with that.

**Comment by rpati12** [comment]
You mean TLS1.3 ? Actually we need tls1.3 and thats the reason I pulled the latest. Anyways, you are right with mbedtls API fragility.

**Comment by lws-team** [comment]
Hm... I bisected it against mbedtls release tags, even the last release of mbedtls-3.5.0 from Oct 4th works fine here, I suggest you try that.

**Comment by rpati12** [comment]
Thanks a lot for your prompt response, I will try that.

**Comment by rpati12** [comment]
I changed mbedtls version as you suggested:

![image](https://github.com/warmcat/libwebsockets/assets/141009171/115eed4a-6aaf-4647-aa07-b7bb0b4143e4)
 
Still I have the same error:

![image](https://github.com/warmcat/libwebsockets/assets/141009171/ddc68bc6-1ea5-4554-a26a-73652f9ca65b)

Following is my LWS configuration:

> cmake -DLWS_WITH_MBEDTLS=1 -DLWS_HAVE_MBEDTLS_NET_SOCKETS=ON -DLWS_MBEDTLS_LIBRARIES="/home/projects/lws/mbedtls/_build/install/lib/libmbedtls.a;/home/projects/lws/mbedtls/_build/install/lib/libmbedcrypto.a;/home/projects/lws/mbedtls/_build/install/lib/libmbedx509.a;/home/projects/lws/mbedtls/_build/install/lib/libp256m.a;/home/projects/lws/mbedtls/_build/install/lib/libeverest.a" -DLWS_MBEDTLS_INCLUDE_DIRS=/home/projects/lws/mbedtls/_build/install/include -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=install -DLWS_WITH_STATIC=ON -DLWS_WITH_SHARED=OFF ..

**Comment by lws-team** [comment]
`-DLWS_HAVE_MBEDTLS_NET_SOCKETS=ON` these thing with `LWS_HAVE_XXX` are discovered by cmake at build time by explicitly testing them against your dependencies, you should not force them.  Mbedtls moved their includes around a while ago and this discovers at build-time if your mbedtls is from before or after that.  If you had to force this, it's suspicious.

You should delete the `build/CMakeCache.txt` and do `make clean` and then rebuild and reinstall on both mbedtls and lws when moving mbedtls version.

You should check the timestamp on the .as to confirm they are what you think it is.

**Comment by rpati12** [comment]
I deleted entire build and install folders from both LWS and mbedtls and reconfigured with following configs and built. Still I have the same issue:

mbedtls config;

cmake -DCMAKE_INSTALL_PREFIX=install ..

LWS config:

cmake -DLWS_WITH_MBEDTLS=1 -DLWS_MBEDTLS_LIBRARIES="/home/projects/lws/mbedtls/_build/install/lib/libmbedtls.a;/home/projects/lws/mbedtls/_build/install/lib/libmbedcrypto.a;/home/projects/lws/mbedtls/_build/install/lib/libmbedx509.a;/home/projects/lws/mbedtls/_build/install/lib/libp256m.a;/home/projects/lws/mbedtls/_build/install/lib/libeverest.a" -DLWS_MBEDTLS_INCLUDE_DIRS=/home/projects/lws/mbedtls/_build/install/include -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=install -DLWS_WITH_STATIC=ON -DLWS_WITH_SHARED=OFF ..

**Comment by lws-team** [comment]
Hum... I got fooled by linking against the wrong mbedtls lib.  To stop that happening any more I added a patch on main that prints the version of mbedtls lws is bound to on startup, and made some paranoid recook sequences for both mbedtls and lws.

```
rm CMakeCache.txt ; cmake .. -DUSE_SHARED_MBEDTLS_LIBRARY=On -DENABLE_TESTING=0 -DCMAKE_INSTALL_PREFIX=/usr && make clean;sudo rm -rf /usr/local/include/mbedtls /usr/include/mbedtls /usr/local/lib64/libmbed* /usr/local/lib/libmbed* /usr/lib64/libmbed* ; make && sudo rm -rf install ; sudo make install && sudo cp -rp install/* /usr
```

and

```
rm CMakeCache.txt ; cmake .. -DLWS_WITH_MBEDTLS=1 -DLWS_STATIC_PIC=1 -DLWS_LINK_TESTAPPS_DYNAMIC=1 && make clean; make -j8 && sudo make install && libwebsockets-test-server -s
```

I studied mbedtls-3.4.0, the release before v3.5.0, it needs some small patches (pushed on main) and then works for tls1.2 without any key error.  The main thing was if you have v1.2, you must do theruntime configure for v1.2 only.   It seems it does not support v1.3 for server  in library/ssl_tls.c ssl_conf_version_check()

```
#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && defined(MBEDTLS_SSL_PROTO_TLS1_3)

    if (mbedtls_ssl_conf_is_hybrid_tls12_tls13(conf)) {
...
        if (conf->endpoint == MBEDTLS_SSL_IS_SERVER) {
            MBEDTLS_SSL_DEBUG_MSG(1, ("TLS 1.3 server is not supported yet."));
            return MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE;
        }
...
```

**Comment by rpati12** [comment]
Ok. So, any plan for you to have the support for the mbedtls 3.5+ support on LWS ? Right now, we might be fine on client side with tls1.3, but in near future we would need it on both sides.

**Comment by lws-team** [comment]
I'm primarily a consultant, if you would like me to work on specific things, you can email me to discuss doing that.

Openssl v1.3 is fine for a long time, AFAIK mbedtls only supported v1.2 when I originally integrated it, I am not sure about its v1.3 status or server capability.

**Comment by rpati12** [comment]
Ahh...ok. Good to know that.

Basically, mbedtls with 1.3 we needed that for constraint embedded devices. I will come back on this topic.

**Comment by lws-team** [comment]
I studied the private key rejection situation that appears at mbedtls-3.5.0, it also rejects Let's Encrypt keys so it's nothing to do with the lws localhost testing keys or their generation.

I bisected the (3000+) patches between 3.4.0 and 3.5.0, I isolated it down to a large merge, and manually chopped that down to this stanza added about 140 patches befofe 3.5.0 release which breaks it

```
diff --git a/library/pkparse.c b/library/pkparse.c
index e1422df771187d7efa536b876614ce184b835b0e..83291c4c7c6035159b76adc829d61aba7540c912 100644
--- a/library/pkparse.c
+++ b/library/pkparse.c
@@ -1417,6 +1417,12 @@ static int pk_parse_key_pkcs8_unencrypted_der(
 #endif /* MBEDTLS_PK_HAVE_ECC_KEYS */
     return MBEDTLS_ERR_PK_UNKNOWN_PK_ALG;

+    end = p + len;
+    if (end != (key + keylen)) {
+        return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PK_KEY_INVALID_FORMAT,
+                                 MBEDTLS_ERR_ASN1_LENGTH_MISMATCH);
+    }
+
     return 0;
 }

```

Basically it's now checking that we ended taking the private key, at the address the end of the private key was supposed to be at in the source buffer now, and for us, the supposed key source buffer is one byte longer, so it doesn't match and ends up with this symptom.

This seems to be heralded in `ChangeLog.d/Switch-pkparse-to-new-pbe-funsctions.txt` by

```
Bugfix
   * mbedtls_pk_parse_key() now rejects trailing garbage in encrypted keys.
```

I made a small patch to use the exact length if mbedtls is >= 3.5.0 and pushed it on main and v4.3-stable, so this should be solved now both for 3.5.0 and mbedtls main.
